From 5b9c6d6db41b9108340c93a5d405ae3e5deee87c Mon Sep 17 00:00:00 2001
From: Ezra Buehler <ezra.buehler@husqvarnagroup.com>
Date: Tue, 15 Oct 2024 20:14:56 +0200
Subject: [PATCH] mips: mt7688: gardena-smart-gateway: Use DM API for flash
 access

Since commit e2e95e5e2542 ("spi: Update speed/mode on change") the
board-specific code for reading/writing factory data from the NOR flash
is problematic, as it (unintentionally) modifies the SPI clock,
also making subsequent NAND flash access slow.

Reading/writing the factory data using the device model API fixes the
issue. For obvious reasons, we chose to modernize the code over simply
adjusting the value of CONFIG_SF_DEFAULT_SPEED.

Signed-off-by: Ezra Buehler <ezra.buehler@husqvarnagroup.com>
Upstream-Status: Pending
---
 board/gardena/smart-gateway-mt7688/board.c | 49 ++++++++--------------
 1 file changed, 18 insertions(+), 31 deletions(-)

diff --git a/board/gardena/smart-gateway-mt7688/board.c b/board/gardena/smart-gateway-mt7688/board.c
index 0cfde91c94c6..bcb8e37b635e 100644
--- a/board/gardena/smart-gateway-mt7688/board.c
+++ b/board/gardena/smart-gateway-mt7688/board.c
@@ -1,3 +1,4 @@
+
 // SPDX-License-Identifier: GPL-2.0+
 /*
  * Copyright (C) 2018 Stefan Roese <sr@denx.de>
@@ -5,6 +6,7 @@
 
 #include <common.h>
 #include <command.h>
+#include <dm/uclass.h>
 #include <env.h>
 #include <env_internal.h>
 #include <init.h>
@@ -61,6 +63,7 @@ int board_early_init_f(void)
 	return 0;
 }
 
+#ifndef CONFIG_SPL_BUILD
 static bool prepare_uuid_var(const char *fd_ptr, const char *env_var_name,
 			     char errorchar)
 {
@@ -89,7 +92,7 @@ static bool prepare_uuid_var(const char *fd_ptr, const char *env_var_name,
 static void factory_data_env_config(void)
 {
 	struct factory_data_values *fd;
-	struct spi_flash *sf;
+	struct udevice *dev;
 	int env_updated = 0;
 	char str[UUID_STR_LEN + 1];	/* Enough for UUID stuff */
 	char *env;
@@ -107,20 +110,16 @@ static void factory_data_env_config(void)
 	/*
 	 * Get values from factory-data area in SPI NOR
 	 */
-	sf = spi_flash_probe(CONFIG_SF_DEFAULT_BUS,
-			     CONFIG_SF_DEFAULT_CS,
-			     CONFIG_SF_DEFAULT_SPEED,
-			     CONFIG_SF_DEFAULT_MODE);
-	if (!sf) {
+	ret = uclass_get_device_by_name(UCLASS_SPI_FLASH, "spi-flash@0", &dev);
+	if (ret) {
 		printf("F-Data:Unable to access SPI NOR flash\n");
 		goto err_free;
 	}
 
-	ret = spi_flash_read(sf, FACTORY_DATA_OFFS, FACTORY_DATA_SIZE,
-			     (void *)buf);
+	ret = spi_flash_read_dm(dev, FACTORY_DATA_OFFS, FACTORY_DATA_SIZE, buf);
 	if (ret) {
 		printf("F-Data:Unable to read factory-data from SPI NOR\n");
-		goto err_spi_flash;
+		goto err_free;
 	}
 
 	fd = (struct factory_data_values *)buf;
@@ -173,9 +172,6 @@ static void factory_data_env_config(void)
 		debug("F-Data:Values match current env values\n");
 	}
 
-err_spi_flash:
-	spi_flash_free(sf);
-
 err_free:
 	free(buf);
 }
@@ -209,7 +205,7 @@ static void copy_or_generate_uuid(char *fd_ptr, const char *env_var_name)
 int do_fd_write(struct cmd_tbl *cmdtp, int flag, int argc, char *const argv[])
 {
 	struct factory_data_values *fd;
-	struct spi_flash *sf;
+	struct udevice *dev;
 	u8 *buf;
 	int ret = CMD_RET_FAILURE;
 
@@ -219,11 +215,8 @@ int do_fd_write(struct cmd_tbl *cmdtp, int flag, int argc, char *const argv[])
 		return CMD_RET_FAILURE;
 	}
 
-	sf = spi_flash_probe(CONFIG_SF_DEFAULT_BUS,
-			     CONFIG_SF_DEFAULT_CS,
-			     CONFIG_SF_DEFAULT_SPEED,
-			     CONFIG_SF_DEFAULT_MODE);
-	if (!sf) {
+	ret = uclass_get_device_by_name(UCLASS_SPI_FLASH, "spi-flash@0", &dev);
+	if (ret) {
 		printf("F-Data:Unable to access SPI NOR flash\n");
 		goto err_free;
 	}
@@ -231,11 +224,10 @@ int do_fd_write(struct cmd_tbl *cmdtp, int flag, int argc, char *const argv[])
 	/* Generate the factory-data struct */
 
 	/* Fist read complete sector into buffer */
-	ret = spi_flash_read(sf, FACTORY_DATA_OFFS, FACTORY_DATA_SECT_SIZE,
-			     (void *)buf);
+	ret = spi_flash_read_dm(dev, FACTORY_DATA_OFFS, FACTORY_DATA_SECT_SIZE, buf);
 	if (ret) {
 		printf("F-Data:spi_flash_read failed (%d)\n", ret);
-		goto err_spi_flash;
+		goto err_free;
 	}
 
 	fd = (struct factory_data_values *)buf;
@@ -272,34 +264,29 @@ int do_fd_write(struct cmd_tbl *cmdtp, int flag, int argc, char *const argv[])
 
 	fd->crc = crc32(0, (u8 *)&fd->magic, FACTORY_DATA_CRC_LEN);
 
-	ret = spi_flash_erase(sf, FACTORY_DATA_OFFS, FACTORY_DATA_SECT_SIZE);
+	ret = spi_flash_erase_dm(dev, FACTORY_DATA_OFFS, FACTORY_DATA_SECT_SIZE);
 	if (ret) {
 		printf("F-Data:spi_flash_erase failed (%d)\n", ret);
-		goto err_spi_flash;
+		goto err_free;
 	}
 
-	ret = spi_flash_write(sf, FACTORY_DATA_OFFS, FACTORY_DATA_SECT_SIZE,
-			      buf);
+	ret = spi_flash_write_dm(dev, FACTORY_DATA_OFFS, FACTORY_DATA_SECT_SIZE, buf);
 	if (ret) {
 		printf("F-Data:spi_flash_write failed (%d)\n", ret);
-		goto err_spi_flash;
+		goto err_free;
 	}
 
 	printf("F-Data:factory-data values written to SPI NOR flash\n");
 
-err_spi_flash:
-	spi_flash_free(sf);
-
 err_free:
 	free(buf);
 
 	return ret;
 }
 
-#ifndef CONFIG_SPL_BUILD
 U_BOOT_CMD(
 	fd_write,	1,	0,	do_fd_write,
 	"Write test factory-data values to SPI NOR",
 	"\n"
 );
-#endif
+#endif /* CONFIG_SPL_BUILD */
-- 
2.43.0

